### ISSUE TITLE
Help with maintenance

### ISSUE THREAD

**Issue by Keats** [Keats/validator#201]
Hey everyone,

As you might have noticed I'm not very reactive with issues in that repository. The main reason is that I'm not (actually never have) used this crate myself so it's not really a priority for me.
I would be happy to promote some new maintainers to handle the new features/fixes they need rather than being the bottleneck. I will still be available if needed, especially for releases etc of course.

**Comment by krojew** [comment]
I might be able to help you. I already maintain 4 crates.

**Comment by Keats** [comment]
That would be great! The main thing to do really is to look at the PRs to see what's missing from them/whether we should close them or not.

**Comment by krojew** [comment]
As far as I can see, most PRs require test fixes and/or rebasing. The only ones, which can be merged now are #140 and #203. I'm not sure #189 is a good idea, since it introduces potential to accidentally ignore fields for validation.

**Comment by Keats** [comment]
I'm not sure #140  makes sense to have in the library itself

**Comment by krojew** [comment]
Domain validation sounds useful, so it might be a good idea to have it alongside the email one.

**Comment by IniterWorker** [comment]
@krojew and @Keats what do you propose? 
- I don't want to switch every time a new outsider comes. So, pick one direction, please.
- @Keats was not happy with a "skip" validator.

**Comment by Keats** [comment]
I think your PR direction is fine, it only allows for _more_ validation since you would get  a compile-time error otherwise. I'm not sure what @krojew means by accidentally ignoring fields for validation.

**Comment by krojew** [comment]
I might have misunderstood how that PR works.

**Comment by krojew** [comment]
What's the plan on moving forward?

**Comment by krojew** [comment]
To amend my question - if you don't want to support the crate, I can take over, since I'm very much interested in keeping it alive.

**Comment by Keats** [comment]
I mostly need help for code reviews and guidance for new features (eg https://github.com/Keats/validator/pull/192), although potentially slow. I'm still there for some reviews/merging/releasing.

I think the current master branch will get a new release this week unless someone takes over https://github.com/Keats/validator/pull/192

**Comment by IniterWorker** [comment]
@Keats could you reviews #189, I will try to get into #192 end of the week. I need more pratice on writting macro.

**Comment by Keats** [comment]
#189 is already merged!

**Comment by IniterWorker** [comment]
I am seeing staled issue. Should we instore a closing issue policy ?

**Comment by Keats** [comment]
I'm not a big fan of stale issues - some old issues are still valid. We should rather go through the issues one by one and close if needed.

**Comment by bradleyess** [comment]
Also happy to help out!

**Comment by Keats** [comment]
Thanks! Any thoughts on https://github.com/Keats/validator/pull/206 and https://github.com/Keats/validator/issues/205?

**Comment by krojew** [comment]
#205 looks quite good - it will greatly improve extensibility and ease maintenance.

**Comment by Keats** [comment]
It would need someone willing to implement it though, it would be a pretty big PR.

**Comment by IniterWorker** [comment]
#170 - We don't have the macro logic to support standard Ref and Cell. Any thoughts about this feature?

**Comment by krojew** [comment]
I think it's an another argument for implementing #205

**Comment by IniterWorker** [comment]
Agree. We should be able to be backward compatible for at least all `validator`'s functions. But, I have some concerns about` extract_custom_validation` and custom functions.

Should we focus on #205?

**Comment by krojew** [comment]
In my opinion, #205 looks promising and should be prioritized.

**Comment by Keats** [comment]
I think with #205 , it might be worth doing a full rewrite of this crate. Probably easier than trying to fit it in and I'm sure the crappy code that was me learning proc macro can be written much better.

**Comment by IniterWorker** [comment]
I started to think about [#205](https://github.com/Keats/validator/issues/205) as a refactoring __without breaking end-user macro-API__.

- Should we have a conservative approach?

However, I don't have too much time these days.

**Comment by Keats** [comment]
I'm not expecting the end-user api to change but all the internals. The issue with a conservative approach is that https://github.com/Keats/validator/pull/192 needs deep refactoring either way, so it's easier/faster to start from scratch imo than bolting on more features

**Comment by DXist** [comment]
I noticed this issue and decided to mention [my PR](https://github.com/Keats/validator/pull/226) here to get potential feedback from active crate users .

I'm interested whether the feature is useful enough to integrate into the validator crate.

**Comment by DXist** [comment]
What do you think about `length_utf16` [validator](https://github.com/Keats/validator/pull/245)?

I think it may be useful to match length constraints in backend and frontend validators - JavaScript returns length in UTF16 characters while `length` validator checks Unicode codepoints.

**Comment by jprochazk** [comment]
Seeing as the general consensus is that a rewrite makes sense, and that large breaking changes are unavoidable, I figured I'd try to start from scratch with a new API. I gave it a shot over the weekend, and the result of this is [garde](https://github.com/jprochazk/garde). I gathered various feature requests that I found here and either implemented them or created an [issue](https://github.com/jprochazk/garde/issues) for them.

It's mostly the same in terms of usage. The core is a `Validate` trait and its `validate` method. The implementation of this trait may be automatically derived.

[Documentation](https://docs.rs/garde/latest/garde/) is a little bare at the moment, so [tests](https://github.com/jprochazk/garde/tree/main/garde_derive_tests/tests) are the best place to look for usage examples.

The major differences are:
- Validation rules are traits, so they may be implemented for custom types. (#205)
- The generated trait accepts a `Context` type, which is passed to each `custom` validator. This allows arbitrary customization to the validation process, and it opens the space up for use-cases like localization of error messages.
- The derive macro supports enums. (#77)
- Added `Valid<T>` and `Unvalidated<T>`, which implement the typestate pattern. (#185)
  `Unvalidated` may be created from arbitrary values, but the only way to create `Valid` is to call `validate` on `Unvalidated`. `Unvalidated` also implements `serde::Deserialize`, so it may be used in [integrations with web frameworks](https://github.com/jprochazk/garde/issues/12).

[Lots of remaining work to do](https://github.com/jprochazk/garde/milestone/1), but as far as I can tell, the only features which are missing for feature parity with `validator` are the `nested` and `required` rules. I'm really interested in seeing what people think!

**Comment by Keats** [comment]
There are some neat ideas that I should definitely take for validator. 

> Rules are responsible for generating error messages, and the resulting errors do not receive params afterwards. This means that sensitive should no longer be necessary, but I might also be totally wrong about this approach, not sure about this!

Sometimes error messages are not generated on the server and you might accidentally log things if there is a default error message (I'm not sure I haven't looked into that in `garde`).

The name `garde` is a tiny bit weird because i associate it with `to keep` first before `to guard` which is both `garder` in French but that's not a big deal.

**Comment by jprochazk** [comment]
> Sometimes error messages are not generated on the server and you might accidentally log things if there is a default error message (I'm not sure I haven't looked into that in `garde`).

`garde` does not log the values at all. The error messages are generated only from the constraint parameters, such as:

```rust
#[derive(Validate)]
struct Test {
  #[garde(range(min=100))]
  field: u64,
}

println!("{:?}", Test { field: 0 }.validate(&()).unwrap_err().flatten());
// prints:
// [("value.field","lower than 10")]
```

> The name `garde` is a tiny bit weird because i associate it with `to keep` first before `to guard` which is both `garder` in French but that's not a big deal.

It seemed appropriate to me, because I typically use validation in a "guard branch", like:
```rust
if let Err(e) = thing.validate() {
  return Err(e);
}
// or just
thing.validate()?;

happy_path();
```
But I don't know any French, so fair enough!

**Comment by Keats** [comment]
> garde does not log the values at all. The error messages are generated only from the constraint parameters

Yeah that's definitely one way to do it and is probably the best way to handle it tbh.

**Comment by IniterWorker** [comment]
> But I don't know any French, so fair enough!

> The name garde is a tiny bit weird because i associate it with to keep first before to guard which is both garder in French but that's not a big deal.

I am a native french speaker, `garde` is good enough. If you say `la garde` in french then it means your guard or a guard watching. And, I think it would fit the application goal protected by watching.

Example from https://www.wordreference.com/fren/garde
- À la boxe, il ne faut jamais baisser sa garde.
- Never drop your guard when boxing.

@jprochazk, I think we should have some community chat to work together.
